### ISSUE TITLE
Extremely slow vector indexing

### ISSUE THREAD

**Issue by orospakr** [tinygrad/tinygrad#5241]
When using vector indexing on a large tensor, performance is something like 100x slower (or worse) than torch. This seems to be the case with both CLANG and METAL device backends.

tinygrad version: 0.9.1 
python version: 3.12.3  
numpy version: 2.0.0
host: macOS 14 arm64

For those who may be interested, this use case came up while going through Karpathy's [bigram predictor exercise](https://youtu.be/PaCmpygFfXo?si=-DKthk2pFPVirvpJ&t=5838). The exercise takes advantage of vector indexing (using `arange`) for building its cost function by referencing output feature row values by index.

```python
#!/usr/bin/env python3

from tinygrad import Tensor, Device, dtypes
import numpy as np

import torch

# this bug does not appear to be particular to device, very slow with both METAL and CLANG.
# Device.DEFAULT = "CLANG"

def main():
    # when using torch, this takes about 2s to start and complete on my m1 laptop.
    # when using tinygrad, it takes over a minute (and sometimes just crashes).
    use_torch = False

    # using data from a np array in order to load this size of real data into
    # torch or tinygrad, rather than a lazily evaluated virtual tensor.
    data = np.zeros((20000, 20000))

    if use_torch:
        huge = torch.tensor(data, dtype=torch.int32)
        indexed = huge[torch.arange(20000), torch.arange(20000)]
        print(indexed[0].sum().item())
    else:
        huge = Tensor(data, dtype=dtypes.int32)
        indexed = huge[Tensor.arange(20000), Tensor.arange(20000)]
        print(indexed[0].sum().item())

if __name__ == "__main__":
    main()
```

**Comment by spullara** [comment]
![profile](https://github.com/user-attachments/assets/4e01d178-c970-4ffd-8a7b-531793356154)

adding a profile for the code generated with py-spy, might be helpful.

**Comment by chenyuxyz** [comment]
i think this is fixed on master. feel free to open a new issue
